server/src/main/java/org/elasticsearch/rest/RestController.java

```diff
@@ -344,6 +345,17 @@ private void tryAllHandlers(final RestRequest request, final RestChannel channel
                         BytesRestResponse.
                             createSimpleErrorResponse(channel, BAD_REQUEST, "multiple values for single-valued header [" + name + "]."));
                     return;
+                } else if (name.equals(Task.TRACE_PARENT)) {
+                    String traceparent = distinctHeaderValues.get(0);
+                    // TODO rely on high performance `TraceContext` from apm-agent-java
+                    // This does little validation of the actual id values, since we only use them to persist in logging for now this
+                    // should be fine. Once we actually start instrumenting through apm-agent-java it will take care of validation.
+                    // Java has a fast path for single character regexes so this should not be as bad as its signature implies
+                    String[] tokens  = traceparent.split("-");
+                    if (tokens.length == 4) {
+                        threadContext.putTransient(Task.TRACE_ID, tokens[1]);
+                    }
+                    threadContext.putHeader(name, String.join(",", distinctHeaderValues));
```

Comment:
Not sure if it's prettier but should be a bit more efficient (no regex and no String[] allocations).
However, it doesn't validate the header at all now.

```suggestion
                } else if (name.equals(Task.TRACE_PARENT)) {
                    String traceparent = distinctHeaderValues.get(0);
                    if (traceparent.length() >= 55) {
                        threadContext.putTransient(Task.TRACE_ID, traceparent.substring(3, 35));
                    }
                    threadContext.putHeader(name, traceparent);
```

Reply:
Thank you will roll with your suggestion!
 
Either way we'd want to start relying on apm-java-agent to take ownership of validation rather than completely reinventing the wheel here.